### ISSUE TITLE
WeKws Roadmap 2.0

### ISSUE THREAD

**Issue by robin1001** [wenet-e2e/wekws#121]
WeKws is a community-driven project and we love your feedback and proposals on where we should be heading.
Feel free to volunteer yourself if you are interested in trying out some items(they do not have to be on the list).

The following items are in 2.0:
[ ] Rubustness, improve the robustness by learning acoustic feature rather than other features of the keywords.
[ ] Various on-device chips support.

**Comment by harryfyodor** [comment]
Hi，Robin. I found that modelscope has opensourced a [model](https://modelscope.cn/models/damo/speech_charctc_kws_phone-xiaoyun/summary) that borrow many codes from wekws. They use CTC as loss function and it seems to work well. I think there are two directions as least worth trying:
* Pretrain in the manner of ASR, and train KWS model with limited data.
* Develop a customizable wake up word system.
Thank you for opensource wekws anyway! It is wonderful!

**Comment by duj12** [comment]
> Hi，Robin. I found that modelscope has opensourced a [model](https://modelscope.cn/models/damo/speech_charctc_kws_phone-xiaoyun/summary) that borrow many codes from wekws. They use CTC as loss function and it seems to work well. I think there are two directions as least worth trying:
> 
> * Pretrain in the manner of ASR, and train KWS model with limited data.
> * Develop a customizable wake up word system.
> 
> Thank you for opensource wekws anyway! It is wonderful!

Hi, I implement this in PR https://github.com/wenet-e2e/wekws/pull/135, of course I borrow a lot of codes from modelscope, too. Hope this will be a good solution. But there is still some thing need to be done, especially runtime code.

**Comment by StuartIanNaylor** [comment]
Hi, Robin. My personal feeling is the KWS we have are good but the manner we train and the datasets we use are not so great.
KWS are usually simple classification models and often the dataset is just a binary of KW and !KW as x2 labels that during training often the accuracy curve is a sign of overfitting as the space between KW and the other classification of anything that is not a KW is huge.
Often a 'noise' classification of non spoken audio increases accuracy as adding classifications gives the model something to train to.
Noise & !KW are non KW catch alls but still distant from the KW of choice 
This is where we are likely shy on datasets as phonetically similar words should create further classifications to hug the KW more tightly and give the model data to train harder on and create a less steep learning curve and negate the overfitting model.

The simple models we have are not the problem its the datasets that are liguistically analysed to create phonetic classification around a KW that are missing.
I have been trying this using the [Multilingual Spoken Words](https://mlcommons.org/en/multilingual-spoken-words) which is just extracted from the Mozilla Common Voice project.
Sadly Common Voice contains huge swaithes of bad data, from wrong labels, bad recordings, alignment and non native language speakers, also the metadata is extremely sparse.

If you had the datasets and also datasets of the device of use then you can make accurate simple and lite KWS so its a bit of a catch-22, but if you picked a model and a device and gave users an option to opt-in then the dataset could be collected as did big data.

**Comment by heibaidaolx123** [comment]
Hi, @robin1001 and @duj12 , any plan for ctc-kws runtime?

**Comment by StuartIanNaylor** [comment]
I just noticed Mining Effective Negative Training Samples for Keyword Spotting ([github](https://github.com/jingyonghou/KWS_Max-pooling_RHE), [paper](https://www.microsoft.com/en-us/research/uploads/prod/2020/04/ICASSP2020_Max_pooling_KWS.pdf))

I have been wondering about a dataset creator and how to select !KW without class imbalance.
I have only had a brief look through the code but is there a dataset creator and also one that implements the above?
